File: sdks/go/qnsp/kms/kms.go
// Package kms is the QNSP KMS client — server-side PQC keys with sign,
// verify, wrap, and unwrap.
package kms

import (
	"context"
	"encoding/base64"
	"fmt"
	"net/http"
	"time"

	"github.com/cuilabs/qnsp-public/sdks/go/qnsp/internal/qnspcore"
)

const pathPrefix = "/kms/v1"

type Client struct {
	base *qnspcore.ServiceClient
}

func New(act *qnspcore.Activator, httpClient *http.Client, timeout time.Duration) *Client {
	return &Client{base: qnspcore.NewServiceClient(act, httpClient, pathPrefix, timeout)}
}

type CreateKeyRequest struct {
	Algorithm string         `json:"algorithm"`
	Purpose   string         `json:"purpose"` // "signing", "encryption", "kem"
	Metadata  map[string]any `json:"metadata,omitempty"`
}

func (c *Client) CreateKey(ctx context.Context, req CreateKeyRequest, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPost, "/keys", req, nil, idempotencyKey)
}

func (c *Client) ListKeys(ctx context.Context, query map[string]string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodGet, "/keys", nil, query, "")
}

func (c *Client) GetKey(ctx context.Context, keyID string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodGet, "/keys/"+keyID, nil, nil, "")
}

func (c *Client) RotateKey(ctx context.Context, keyID string, idempotencyKey string) (map[string]any, error) {
	return c.base.Do(ctx, http.MethodPost, "/keys/"+keyID+"/rotate", nil, nil, idempotencyKey)
}

func (c *Client) DeleteKey(ctx context.Context, keyID string) error {
	_, err := c.base.Do(ctx, http.MethodDelete, "/keys/"+keyID, nil, nil, "")
	return err
}

func (c *Client) Sign(ctx context.Context, keyID string, data []byte, idempotencyKey string) ([]byte, error) {
	body := map[string]any{"dataB64": base64.StdEncoding.EncodeToString(data)}
	resp, err := c.base.Do(ctx, http.MethodPost, "/keys/"+keyID+"/sign", body, nil, idempotencyKey)
	if err != nil {
		return nil, err
	}
	sigB64, ok := resp["signatureB64"].(string)
	if !ok || sigB64 == "" {
		return nil, fmt.Errorf("qnsp: kms.Sign: missing signatureB64 in response")
	}
	return base64.StdEncoding.DecodeString(sigB64)
}

func (c *Client) Verify(ctx context.Context, keyID string, data, signature []byte) (bool, error) {
	body := map[string]any{
		"dataB64":      base64.StdEncoding.EncodeToString(data),
		"signatureB64": base64.StdEncoding.EncodeToString(signature),
	}
	resp, err := c.base.Do(ctx, http.MethodPost, "/keys/"+keyID+"/verify", body, nil, "")
	if err != nil {
		return false, err
	}
	valid, _ := resp["valid"].(bool)
	return valid, nil
}

func (c *Client) Wrap(ctx context.Context, keyID string, plaintext []byte, idempotencyKey string) ([]byte, error) {
	body := map[string]any{"plaintextB64": base64.StdEncoding.EncodeToString(plaintext)}
	resp, err := c.base.Do(ctx, http.MethodPost, "/keys/"+keyID+"/wrap", body, nil, idempotencyKey)
	if err != nil {
		return nil, err
	}
	ctB64, ok := resp["ciphertextB64"].(string)
	if !ok || ctB64 == "" {
		return nil, fmt.Errorf("qnsp: kms.Wrap: missing ciphertextB64 in response")
	}
	return base64.StdEncoding.DecodeString(ctB64)
}

func (c *Client) Unwrap(ctx context.Context, keyID string, ciphertext []byte, idempotencyKey string) ([]byte, error) {
	body := map[string]any{"ciphertextB64": base64.StdEncoding.EncodeToString(ciphertext)}
	resp, err := c.base.Do(ctx, http.MethodPost, "/keys/"+keyID+"/unwrap", body, nil, idempotencyKey)
	if err != nil {
		return nil, err
	}
	ptB64, ok := resp["plaintextB64"].(string)
	if !ok || ptB64 == "" {
		return nil, fmt.Errorf("qnsp: kms.Unwrap: missing plaintextB64 in response")
	}
	return base64.StdEncoding.DecodeString(ptB64)
}

File: packages/qnsp/src/kms.ts
/**
 * QNSP KMS — server-side PQC keys with sign, verify, wrap, and unwrap.
 * Wraps `apps/kms-service` (`/kms/v1`).
 */

import type { Internal, RequestOptions } from "./_internal.js";
import { QnspApiError } from "./errors.js";

const PATH_PREFIX = "/kms/v1";

export interface CreateKeyRequest {
	readonly algorithm: string;
	readonly purpose: string; // "signing" | "encryption" | "kem"
	readonly metadata?: Record<string, unknown>;
}

interface SignResponse {
	readonly signatureB64?: string;
}

interface VerifyResponse {
	readonly valid?: boolean;
}

interface WrapResponse {
	readonly ciphertextB64?: string;
}

interface UnwrapResponse {
	readonly plaintextB64?: string;
}

export class KmsClient {
	constructor(private readonly internal: Internal) {}

	createKey(req: CreateKeyRequest, opts?: Pick<RequestOptions, "idempotencyKey">) {
		return this.internal.request("POST", `${PATH_PREFIX}/keys`, req, opts);
	}

	listKeys(query?: RequestOptions["query"]) {
		return this.internal.request("GET", `${PATH_PREFIX}/keys`, undefined, { query });
	}

	getKey(keyId: string) {
		return this.internal.request("GET", `${PATH_PREFIX}/keys/${keyId}`);
	}

	rotateKey(keyId: string, opts?: Pick<RequestOptions, "idempotencyKey">) {
		return this.internal.request("POST", `${PATH_PREFIX}/keys/${keyId}/rotate`, undefined, opts);
	}

	deleteKey(keyId: string): Promise<Record<string, unknown>> {
		return this.internal.request("DELETE", `${PATH_PREFIX}/keys/${keyId}`);
	}

	async sign(
		keyId: string,
		data: Uint8Array,
		opts?: Pick<RequestOptions, "idempotencyKey">,
	): Promise<Uint8Array> {
		const body = { dataB64: encodeB64(data) };
		const resp = await this.internal.request<SignResponse>(
			"POST",
			`${PATH_PREFIX}/keys/${keyId}/sign`,
			body,
			opts,
		);
		if (!resp.signatureB64) {
			throw new QnspApiError("kms.sign: response missing signatureB64", 200);
		}
		return decodeB64(resp.signatureB64);
	}

	async verify(keyId: string, data: Uint8Array, signature: Uint8Array): Promise<boolean> {
		const body = { dataB64: encodeB64(data), signatureB64: encodeB64(signature) };
		const resp = await this.internal.request<VerifyResponse>(
			"POST",
			`${PATH_PREFIX}/keys/${keyId}/verify`,
			body,
		);
		return resp.valid === true;
	}

	async wrap(
		keyId: string,
		plaintext: Uint8Array,
		opts?: Pick<RequestOptions, "idempotencyKey">,
	): Promise<Uint8Array> {
		const body = { plaintextB64: encodeB64(plaintext) };
		const resp = await this.internal.request<WrapResponse>(
			"POST",
			`${PATH_PREFIX}/keys/${keyId}/wrap`,
			body,
			opts,
		);
		if (!resp.ciphertextB64) {
			throw new QnspApiError("kms.wrap: response missing ciphertextB64", 200);
		}
		return decodeB64(resp.ciphertextB64);
	}

	async unwrap(
		keyId: string,
		ciphertext: Uint8Array,
		opts?: Pick<RequestOptions, "idempotencyKey">,
	): Promise<Uint8Array> {
		const body = { ciphertextB64: encodeB64(ciphertext) };
		const resp = await this.internal.request<UnwrapResponse>(
			"POST",
			`${PATH_PREFIX}/keys/${keyId}/unwrap`,
			body,
			opts,
		);
		if (!resp.plaintextB64) {
			throw new QnspApiError("kms.unwrap: response missing plaintextB64", 200);
		}
		return decodeB64(resp.plaintextB64);
	}
}

function encodeB64(bytes: Uint8Array): string {
	return Buffer.from(bytes).toString("base64");
}

function decodeB64(b64: string): Uint8Array {
	return new Uint8Array(Buffer.from(b64, "base64"));
}
